File: source/native/src/expression/Context.cpp
#include <Array.hpp>
#include "Context.hpp"
#include "Expression.hpp"

Value Context::get(string name)
{
    return Value();
}

void Context::assign(string name, Value value)
{
}

Value Context::call(string name, vector<shared_ptr<Expression>> arguments)
{
    return Value();
}

shared_ptr<Context> Context::redirect(string name)
{
    return NULL;
}

ObjectContext::ObjectContext(Variant object_)
{
    if (object_.get_type() != Variant::OBJECT) {
        cerr << "Invalid context" << endl;
        object = Object::_new();
        return;
    }

    object = Object::___get_from_variant(object_);
}

Value ObjectContext::get(string name)
{
    if (!object->has_method("get_context_variable")) {
        cerr << "context: godot object has no method get_context_variable" << endl;
        return Value();
    }

    Array godotArguments;
    godotArguments.push_back(name.c_str());
    Variant godotResult = object->callv("get_context_variable", godotArguments);

    return Value(godotResult);
}

void ObjectContext::assign(string name, Value value)
{
    if (!object->has_method("set_context_variable")) {
        cerr << "context: godot object has no method set_context_variable" << endl;
        return;
    }

    Array godotArguments;
    godotArguments.push_back(name.c_str());
    godotArguments.push_back(value);

    object->callv("set_context_variable", godotArguments);
}

shared_ptr<Context> ObjectContext::redirect(string name)
{
    if (!object->has_method("redirect_context")) {
        cerr << "context: godot object has no method redirect_context" << endl;
        return NULL;
    }

    Array godotArguments;
    godotArguments.push_back(name.c_str());
    Variant godotResult = object->callv("redirect_context", godotArguments);

    if (godotResult.get_type() == Variant::NIL) {
        return NULL;
    }

    return shared_ptr<Context>(new ObjectContext(godotResult));
}

Value ObjectContext::call(string name, vector<shared_ptr<Expression>> arguments)
{
    if ("cond" == name) {
        if (arguments.size() == 3) {
            Value cond = arguments[0]->evaluate(this);
            if  (cond.isBottom()) {
                return Value();
            }
            if (cond) {
                return arguments[1]->evaluate(this);
            }
            return arguments[2]->evaluate(this);
        }

        cerr << "context: invalid argument count for cond function: " << arguments.size() << endl;
        return Value();
    }

    if ("ifelse" == name) {
        if (arguments.size() == 3) {
            Value cond = arguments[0]->evaluate(this);
            Value trueValue = arguments[1]->evaluate(this);
            Value falseValue = arguments[2]->evaluate(this);
            if (cond.isBottom()) {
                return Value();
            }
            return cond ? trueValue : falseValue;
        }

        cerr << "context: invalid argument count for ifelse function: " << arguments.size() << endl;
        return Value();
    }

    if (!object->has_method("call_context_function")) {
        cerr << "context: godot object has no method call_context_function" << endl;
        return Value();
    }

    Array innerArguments;

    for (int i = 0; i < arguments.size(); i++) {
        innerArguments.push_back(arguments[i]->evaluate(this));
    }

    Array godotArguments;
    godotArguments.push_back(name.c_str());
    godotArguments.push_back(innerArguments);

    Variant godotResult = object->callv("call_context_function", godotArguments);

    return Value(godotResult);
}

File: source/native/src/expressions/context.rs
use crate::expressions::expression::Expression;
use crate::expressions::value::Value;
use gdnative::prelude::*;
use std::rc::Rc;

pub struct Context {
    object: Variant,
}

impl Context {
    pub fn new(object: Variant) -> Context {
        Context { object }
    }

    pub fn godot_call(&self, name: &str, arguments: &[Variant]) -> Option<Variant> {
        let object_ref = self.object.try_to_object::<Object>();

        if object_ref.is_none() {
            godot_print!("invalid_context");
            return Option::None;
        }

        let object = unsafe { object_ref.expect("invalid object ref").assume_safe() };
        let has_method = object.has_method(name);

        if !has_method {
            godot_print!("method_not_found: {}", name);
            return Option::None;
        }

        unsafe { object.call(name, arguments) }.into()
    }

    pub fn get(&self, name: &str) -> Value {
        let result = self.godot_call("get_context_variable", &[name.to_variant()]);

        match result {
            Option::Some(x) => x.into(),
            Option::None => Value::BottomValue,
        }
    }

    pub fn call(&self, name: &str, expressions: &[Rc<Expression>]) -> Value {
        if "cond" == name {
            if expressions.len() == 3 {
                let cond = expressions[0].evaluate(&self);
                if cond.is_bottom() {
                    return Value::BottomValue;
                }
                if cond.to_bool() {
                    return expressions[1].evaluate(&self);
                }
                return expressions[2].evaluate(&self);
            }

            godot_print!(
                "context: invalid argument count for cond function: {}",
                expressions.len()
            );

            return Value::BottomValue;
        }

        if "ifelse" == name {
            if expressions.len() == 3 {
                let cond = expressions[0].evaluate(&self);
                let true_value = expressions[1].evaluate(&self);
                let false_value = expressions[2].evaluate(&self);
                if cond.is_bottom() {
                    return Value::BottomValue;
                }
                return if cond.to_bool() {
                    true_value
                } else {
                    false_value
                };
            }

            godot_print!(
                "context: invalid argument count for ifelse function: {}",
                expressions.len()
            );

            return Value::BottomValue;
        }

        let mut arguments: Vec<Variant> = Vec::new();
        for it in expressions.iter() {
            arguments.push(it.evaluate(&self).to_variant());
        }
        let result = self.godot_call(
            "call_context_function",
            &[name.to_variant(), arguments.to_variant()],
        );

        match result {
            Option::Some(x) => x.into(),
            Option::None => Value::BottomValue,
        }
    }

    pub fn assign(&self, name: &str, value: &Value) {
        self.godot_call(
            "set_context_variable",
            &[name.to_variant(), value.to_variant()],
        );
    }

    pub fn redirect(&self, name: &str) -> Option<Context> {
        let result = self.godot_call("redirect_context", &[name.to_variant()]);

        match result {
            Option::Some(variant) => Option::Some(Context::new(variant)),
            _ => Option::None,
        }
    }
}
